File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/application/usecases/GiftCardRefunding.java
package it.giannibombelli.wsc2026.giftcard.application.usecases;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.application.UseCase;
import it.giannibombelli.wsc2026.giftcard.application.commands.RefundGiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.events.GiftCardRefunded;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import it.giannibombelli.wsc2026.giftcard.domain.ports.GiftCardRepository;

import static java.util.Objects.requireNonNull;

public class GiftCardRefunding implements UseCase<RefundGiftCard, GiftCardRefunded> {
    private final GiftCardRepository giftCardRepository;

    public GiftCardRefunding(GiftCardRepository giftCardRepository) {
        this.giftCardRepository = requireNonNull(giftCardRepository);
    }

    @Override
    public GiftCardRefunded invoke(RefundGiftCard cmd) {
        Require.requireArgument(cmd, "command");
        GiftCardId giftCardId = cmd.aggregateId();

        GiftCard giftCard = giftCardRepository.findById(giftCardId)
            .orElseThrow(() -> new IllegalStateException("GiftCard not found for credit: " + giftCardId));

        GiftCardRefunded refunded = giftCard.refund(cmd.amount());

        giftCardRepository.save(giftCard);
        return refunded;
    }
}

File: typescript/src/giftcard/application/usecases/giftCardRefunding.ts
// Caso d'uso per il rimborso di una gift card.

import type { UseCase } from '@/common/application/usecase.js';
import type { GiftCardRefunded } from '../../domain/events/giftCardRefunded.js';
import { GiftCardId } from '../../domain/giftcard/giftCardId.js';
import type { GiftCardRepository } from '../../domain/ports/giftCardRepository.js';
import { RefundGiftCard } from '../commands/refundGiftCard.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class GiftCardRefunding implements UseCase<RefundGiftCard, GiftCardRefunded> {
  private readonly _giftCardRepository: GiftCardRepository;

  constructor(giftCardRepository: GiftCardRepository) {
    requireDependency(giftCardRepository, "giftCardRepository");
    this._giftCardRepository = giftCardRepository;
  }

  invoke(command: RefundGiftCard): GiftCardRefunded {
    requireArgument(command, 'command');

    const giftCardId: GiftCardId = command.aggregateId;
    const giftCard = this._giftCardRepository.findById(giftCardId);
    if (giftCard === null) {
      throw new Error(`GiftCard not found for refund: ${giftCardId.value}`);
    }

    const refunded = giftCard.refund(command.amount);

    this._giftCardRepository.save(giftCard);
    return refunded;
  }
}
